text/0002-sarif-result-schema.md

```diff
@@ -0,0 +1,482 @@
+     *
+     * CHECKUP: if we wanted to keep concepts of Result type `SummaryResult | MultiValueResult | LookupValueResult;`, we can store those here, as well as the dataSummary and data for each Result
+     */
+    message: Message;
+
+    /**
+     * A positive integer specifying the number of times this logically unique result was observed in this run.
+     */
+    occurrenceCount?: number;
+
+    /**
+     * A number representing the priority or importance of the result.
+     */
+    rank?: number;
+
+    /**
+     * The stable, unique identifier of the rule, if any, to which this notification is relevant. This member can be
+     * used to retrieve rule metadata from the rules dictionary, if it exists.
+     */
+    ruleId?: string;
+
+    /**
+     * The index within the tool component rules array of the rule object associated with this result.
+     */
+    ruleIndex?: number;
+
+    /**
+     * Key/value pairs that provide additional information about the result.
+     * 
+     */
+    properties?: PropertyBag;
+}
+
+/**
+ * Encapsulates a message intended to be read by the end user.
+ */
+export interface Message {
+    /**
+     * An array of strings to substitute into the message string.
+     */
+    arguments?: string[];
+
+    /**
+     * The identifier for this message.
+     */
+    id?: string;
+
+    /**
+     * A Markdown message string.
+     */
+    markdown?: string;
+
+    /**
+     * A plain text message string.
+     */
+    text?: string;
+
+    /**
+     * Key/value pairs that provide additional information about the message.
+     */
+    properties?: PropertyBag;
+}
+
+/**
+ * A proposed fix for the problem represented by a result object. A fix specifies a set of artifacts to modify. For
+ * each artifact, it specifies a set of bytes to remove, and provides a set of new bytes to replace them.
+ */
+export interface Fix {
+    /**
+     * One or more artifact changes that comprise a fix for a result.
+     */
+    artifactChanges: ArtifactChange[];
+
+    /**
+     * A message that describes the proposed fix, enabling viewers to present the proposed change to an end user.
+     */
+    description?: Message;
+
+    /**
+     * Key/value pairs that provide additional information about the fix.
+     */
+    properties?: PropertyBag;
+}
+
+
+```
+Note: there are type definitions for SARIF [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/289ba098a611b80ee03ed72a2feee41a593fe765/types/sarif/index.d.ts). These types are being used by a number of the high profile tools I linked above, so I view them as very reliable, even though they are third party.
+
+## Open questions: 
+1) Where is it best to store TaskErrors? Options are: properties on `Invocation`, properties on `Run`. I don't see a more specific place in SARIF schema to put nonterminal errors
+2) Are there concerns with storing `actions` as `fixes` per each `Result`, instead of aggregating all `actions` a level up  (on the `Run` itself)?
```

Comment:
This seems like a good change, and will allow us to align with SERIF's format.

Reply:
@scalvert - in trying to implement this, I realized that `Fix` is not really an appropriate construct to cleanly shove `Actions` into, as it requires [ArtifactChange](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/289ba098a611b80ee03ed72a2feee41a593fe765/types/sarif/index.d.ts#L201) and [ArtifactLocation](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/289ba098a611b80ee03ed72a2feee41a593fe765/types/sarif/index.d.ts#L247) which _could_ apply to some actions, but not cleanly to all at this time. I am going to leave `Actions` in the `PropertyBag` for initial migration, and then re-evaluate if we want change how we think about Actions to align closer with the SARIF construct of Fixes